lxd/cluster/recover.go

```diff
@@ -160,51 +205,214 @@ func Reconfigure(database *db.Node, raftNodes []db.RaftNode) error {
 	if localAddress != info.Address {
 		err := updateLocalAddress(database, localAddress)
 		if err != nil {
-			return err
+			return "", err
 		}
 	}
 
 	dir := filepath.Join(database.Dir(), "global")
 	// Replace cluster configuration in dqlite.
 	err = dqlite.ReconfigureMembershipExt(dir, nodes)
 	if err != nil {
-		return fmt.Errorf("Failed to recover database state: %w", err)
+		return "", fmt.Errorf("Failed to recover database state: %w", err)
 	}
 
 	// Replace cluster configuration in local raft_nodes database.
 	err = database.Transaction(context.TODO(), func(ctx context.Context, tx *db.NodeTx) error {
 		return tx.ReplaceRaftNodes(raftNodes)
 	})
+	if err != nil {
+		return "", err
+	}
+
+	tarballPath, err := writeRecoveryTarball(database.Dir(), raftNodes)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create recovery tarball: copy db manually; %w", err)
+	}
+
+	err = writeGlobalNodesPatch(database, raftNodes)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create global db patch for cluster recover: %w", err)
+	}
+
+	return tarballPath, nil
+}
+
+// Create a tarball of the global database dir to be copied to all other
+// remaining cluster members.
+func writeRecoveryTarball(databaseDir string, raftNodes []db.RaftNode) (string, error) {
+	tarballPath := filepath.Join(databaseDir, RecoveryTarballName)
+
+	tarball, err := os.Create(tarballPath)
+	if err != nil {
+		return "", err
+	}
+
+	gzWriter := gzip.NewWriter(tarball)
+	tarWriter := tar.NewWriter(gzWriter)
+
+	globalDBDirFS := os.DirFS(filepath.Join(databaseDir, "global"))
+
+	err = tarWriter.AddFS(globalDBDirFS)
+	if err != nil {
+		return "", err
+	}
+
+	raftNodesYaml, err := yaml.Marshal(raftNodes)
+	if err != nil {
+		return "", err
+	}
+
+	raftNodesHeader := tar.Header{
+		Typeflag: tar.TypeReg,
+		Name:     raftNodesFilename,
+		Size:     int64(len(raftNodesYaml)),
+		Mode:     0o644,
+		Uid:      0,
+		Gid:      0,
+		Format:   tar.FormatPAX,
+	}
+
+	err = tarWriter.WriteHeader(&raftNodesHeader)
+	if err != nil {
+		return "", err
+	}
+
+	written, err := tarWriter.Write(raftNodesYaml)
+	if err != nil {
+		return "", err
+	}
+
+	if written != len(raftNodesYaml) {
+		return "", fmt.Errorf("Wrote %d bytes but expected to write %d", written, len(raftNodesYaml))
+	}
+
+	err = tarWriter.Close()
+	if err != nil {
+		return "", err
+	}
+
+	err = gzWriter.Close()
+	if err != nil {
+		return "", err
+	}
+
+	err = tarball.Close()
+	if err != nil {
+		return "", err
+	}
+
+	return tarballPath, nil
+}
+
+// MaybeUnpackRecoveryTarball checks for the existence of recovery_db.tar.gz in
+// the database dir. If it exists, it unpacks the tarball, replaces the global
+// database, and updates the local database with any changed addresses.
+func MaybeUnpackRecoveryTarball(database *db.Node) error {
```

Comment:
`shared.PathExists` at the call site instead, and pass the tarball path into the func instead of creating it here (separation of concerns).